editor/src/components/editor/store/dispatch-strategies.tsx

```diff
@@ -0,0 +1,600 @@
+        patchedEditorState: newEditorState,
+        newStrategyState: result.strategyState,
+      }
+    }
+  }
+}
+
+export function handleStrategies(
+  dispatchedActions: readonly EditorAction[],
+  storedState: EditorStoreFull,
+  result: InnerDispatchResult,
+  oldDerivedState: DerivedState,
+): HandleStrategiesResult & { patchedDerivedState: DerivedState } {
+  const { unpatchedEditorState, patchedEditorState, newStrategyState } = handleStrategiesInner(
+    dispatchedActions,
+    storedState,
+    result,
+  )
+
+  const patchedEditorWithMetadata: EditorState = {
+    ...patchedEditorState,
+    jsxMetadata:
+      patchedEditorState.canvas.interactionSession?.metadata ?? patchedEditorState.jsxMetadata,
+  }
+
+  const patchedDerivedState = deriveState(patchedEditorWithMetadata, oldDerivedState, 'patched')
+
+  return {
+    unpatchedEditorState,
+    patchedEditorState: patchedEditorWithMetadata,
+    patchedDerivedState,
+    newStrategyState: newStrategyState,
+  }
+}
+
+function handleStrategiesInner(
+  dispatchedActions: readonly EditorAction[],
+  storedState: EditorStoreFull,
+  result: InnerDispatchResult,
+): HandleStrategiesResult {
+  const makeChangesPermanent = dispatchedActions.some(shouldApplyClearInteractionSessionResult)
+  const cancelInteraction =
+    dispatchedActions.some(isClearInteractionSession) && !makeChangesPermanent
+  if (storedState.unpatchedEditor.canvas.interactionSession == null) {
+    if (result.unpatchedEditor.canvas.interactionSession == null) {
+      return {
+        unpatchedEditorState: result.unpatchedEditor,
+        patchedEditorState: result.unpatchedEditor,
+        newStrategyState: result.strategyState,
+      }
+    } else {
+      return interactionStart(storedState, result)
+    }
+  } else {
+    if (cancelInteraction) {
+      return interactionCancel(storedState, result)
+    } else if (makeChangesPermanent) {
+      return interactionFinished(storedState, result)
+    } else {
+      const interactionHardResetNeeded =
+        hasModifiersChanged(
+          storedState.unpatchedEditor.canvas.interactionSession?.interactionData ?? null,
+          result.unpatchedEditor.canvas.interactionSession?.interactionData ?? null,
+        ) || result.strategyState.currentStrategy == null // TODO: do we really need the currentStrategy == null part?
+      if (interactionHardResetNeeded) {
+        return interactionHardReset(storedState, result)
+      } else {
+        if (result.unpatchedEditor.canvas.interactionSession?.userPreferredStrategy != null) {
+          const userChangedStrategy =
+            result.unpatchedEditor.canvas.interactionSession?.userPreferredStrategy !=
+            storedState.unpatchedEditor.canvas.interactionSession?.userPreferredStrategy
+          if (userChangedStrategy) {
+            return interactionUserChangedStrategy(storedState, result)
+          }
+        }
+
+        const strategy = findCanvasStrategyFromDispatchResult(result)
```

Comment:
This is probably too big task to include in this PR, but I don't like that we have to find the canvas strategy to decide whether there is a strategy change, and then inside `interactionStrategyChangeStacked` we find the canvas strategy again. It seems this big `if` here to handle the different interaction strategy cases should be _after_ finding the canvas strategy, and not before. That would also decrease the code duplication between the different `interaction*` functions.
Note that this will be even worse after we actually apply the accumulated commands before finding the strategy, even more code will has to be duplicated.
I think it would make sense to refactor this after the PR is merged

Reply:
done in 6ae9ccf2139367612c61e0c74a73915ecd254138